modin/pandas/test/utils.py

```diff
@@ -573,7 +584,93 @@ def assert_empty_frame_equal(df1, df2):
+    """
+    Assert that all of the objs give the same boolean result for the passed condition (either all True or all False).
+
+    Parameters
+    ----------
+    condition : callable(obj) -> bool
+        Condition to run on the passed objects.
+    *objs :
+        Objects to pass to the condition.
+
+    Returns
+    -------
+    bool
+        Result of the condition.
+    """
+    results = [condition(obj) for obj in objs]
+    if len(results) < 2:
+        return results[0] if len(results) else None
+
+    assert all(results[0] == res for res in results[1:])
+    return results[0]
+
+
+def _maybe_cast_to_pandas_dtype(dtype):
+    """Cast passed `dtype` to according pandas dtype if needed for the sake of equality comparison."""
+    # If we're running in a cloud mode then all the numpy types are substituted with a proxy net-reference,
+    # Such dtypes won't pass equality check with pandas dtypes thus manually converting them to a pure numpy
+    if "netref" in str(type(dtype)):
+        return np.dtype(dtype.name)
+    return dtype
+
+
+def assert_dtypes_equal(df1, df2):
+    """
+    Assert that the two passed DataFrame/Series objects have equal dtypes.
+
+    The function doesn't require that the dtypes are identical, it has the following reliefs:
+        1. The dtypes are not required to be in the same order
+           (e.g. {"col1": int, "col2": float} == {"col2": float, "col1": int})
+        2. The dtypes are only required to be in the same class
+           (e.g. both numerical, both categorical, etc...)
+
+    Parameters
+    ----------
+    df1 : DataFrame or Series
+    df2 : DataFrame or Series
+    """
+    if not isinstance(
+        df1, (pandas.Series, pd.Series, pandas.DataFrame, pd.DataFrame)
+    ) or not isinstance(
+        df2, (pandas.Series, pd.Series, pandas.DataFrame, pd.DataFrame)
+    ):
+        return
+
+    if isinstance(df1.dtypes, (pandas.Series, pd.Series)):
+        dtypes1 = df1.dtypes
+        dtypes2 = df2.dtypes
+    else:
+        # Case when `dtypes` is a scalar
+        dtypes1 = pandas.Series({"col": df1.dtypes})
+        dtypes2 = pandas.Series({"col": df2.dtypes})
+
+    # Don't require for dtypes to be in the same order
+    assert len(dtypes1.index.difference(dtypes2.index)) == 0
+    assert len(dtypes1) == len(dtypes2)
+
+    dtype_comparators = (
+        is_numeric_dtype,
+        lambda obj: is_object_dtype(obj) or is_string_dtype(obj),
+        is_bool_dtype,
+        is_categorical_dtype,
+        is_datetime64_any_dtype,
+        is_timedelta64_dtype,
+        is_period_dtype,
+    )
+
+    for col in dtypes1.keys():
+        type1, type2 = map(_maybe_cast_to_pandas_dtype, (dtypes1[col], dtypes2[col]))
+        for comparator in dtype_comparators:
+            if assert_all_act_same(comparator, type1, type2):
+                # We met a dtype that both types satisfy, so we can stop iterating
+                # over comparators and compare next dtypes
+                break
```

Comment:
No, here we're iterating over a list of `is_some_dtype()` functions and as soon as we meet one that returns True we can stop iterating because we found a proper dtype checker.

Reply:
I assumed that we have to check all dtypes of the dfs, but not only one, which matches both. Isn't that the case?